src/matrixrtc/MatrixRTCSession.ts

```diff
@@ -0,0 +1,385 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { logger } from "../logger";
+import { TypedEventEmitter } from "../models/typed-event-emitter";
+import { EventTimeline } from "../models/event-timeline";
+import { Room } from "../models/room";
+import { MatrixClient } from "../client";
+import { EventType } from "../@types/event";
+import { CallMembership, CallMembershipData } from "./CallMembership";
+import { Focus } from "./focus";
+import { MatrixEvent } from "../matrix";
+
+const MEMBERSHIP_EXPIRY_TIME = 60 * 60 * 1000;
+const MEMBER_EVENT_CHECK_PERIOD = 2 * 60 * 1000; // How often we check to see if we need to re-send our member event
+const CALL_MEMBER_EVENT_RETRY_DELAY_MIN = 3000;
+
+export enum MatrixRTCSessionEvent {
+    // A member joined, left, or updated a proprty of their membership
+    MembershipsChanged = "memberships_changed",
+    // We joined or left the session (our own local idea of whether we are joined, separate from MembershipsChanged)
```

Comment:
```suggestion
    // We joined or left the session (our own local idea of whether we are joined, separate from MembershipsChanged).
```

OTOH: Do we want to differentiate between them? I.e. why does not membership suffice to designate our participation in a call?

Reply:
I've added to the comment a bit: I also wondered this but it felt useful to distinguish whether we intend to be joined to the call and whether our membership event has actually gone through, and this seems like an easier, simpler way of doing local echo.